### PR TITLE
[AP-578] Commit offset from state file and not from the consumed message

### PR DESCRIPTION
**Pull request by koszti**
Currently, tap-kafka sends a commit message back to kafka as soon as it received a message. This can lead to data loss should anything go wrong during the loading into target.

The correct thing is to send the acknowledgement only when the data has been loaded. PipelineWise updates `state.json` when new data loaded into target. This PR reads the state.json periodically and commit the offset and partition safely.

The code had to be refactored a bit to be more testable. Also this PR adds a bunch of new unit tests to check bookmarks, consumed and produced kafka and singer messages and the commit timing.

### CODE REVIEW

File: tap_kafka/sync.py [hunk 1]
```diff
@@ -93,17 +112,43 @@ def sync_stream(kafka_config, stream, state, consumer):
...
 
-    singer.write_message(singer.StateMessage(value=copy.deepcopy(state)))
+        # Every COMMIT_INTERVAL, commit the latest offset from the state_file
+        if not commit_timestamp or \
+                datetime.datetime.utcnow() >= (commit_timestamp + datetime.timedelta(seconds=COMMIT_INTERVAL)):
+            # Read the state from disk, maybe a target connector updated it in the meantime
+            args = fn_get_args()
+            state = args.state or {}
+
+            # Commit the kafka offset
```
Comment (Samira-El) [inline tap_kafka/sync.py]: Is it possible to have an offset that is not None and not have a state?

**Comment by Samira-El** [comment]
Looks good to me, just want to know if it's possible to have e2e or integration tests using a kafka docker image as the source?

**Comment by koszti** [comment]
yes it's possible to add this for example https://github.com/wurstmeister/kafka-docker but it a relatively complex. Where e2e/integration with real kafka should be? MySQL and Postgres taps don't have it.

Would that be fine if I try to add kafka e2e with a docker image in the main pipelinewise?

At the moment this repo is using a mocked KafkaConsumer class at https://github.com/transferwise/pipelinewise-tap-kafka/blob/f1126c1aa9e93247ac52b30835b62c2e8c60ef95/tests/helper/kafka_consumer_mock.py

end to end

**Comment by Samira-El** [comment]
Tap-mysql has a mariadb docker image set in place for testing :)

We can add then in PPW instead.

**Comment by koszti** [comment]
wow, you're right, sorry about that. tap-mysql really does have docker image. I didn't know and completely missed that. Nice!

**Pull request merged**